### PR TITLE
Adding Node module for snowboy

### PR DESCRIPTION
**Pull request by evancohen**
## Authors
- @nekuz0r
- @evancohen

## Info
This PR adds an npm package that uses [`node-gyp`](https://github.com/nodejs/node-gyp) with [`Nan`](https://github.com/nodejs/nan) to create a JavaScript interface between the prebuilt snowboy native hotword detection and the Node runtime. 

Using [`node-pre-gyp`](https://github.com/mapbox/node-pre-gyp), we are able to host pre-compiled versions of the native addon for OSX, Linux x64, and the Raspberry Pi (arm), which makes using the package super easy!

More info: #4 

Installation of this package uses a platform independent command:
```
npm install --save snowboy
```
## Usage
There are two [examples](https://github.com/evancohen/snowboy/tree/devel/examples/Node) provided in this PR for streaming from a file and directly from a microphone. The short version of this is:

``` javascript
const record = require('node-record-lpcm16');
const {Detector, Models} = require('snowboy');

const models = new Models();

models.add({file: 'resources/snowboy.umdl', sensitivity: '0.5', hotwords : 'snowboy'});

const detector = new Detector({
  resource: "resources/common.res",
  models: models,
  audioGain: 2.0
});

detector.on('hotword', function (index, hotword) {
  console.log('hotword', index, hotword);
});

const mic = record.start({
  threshold: 0,
  verbose: true
});

mic.pipe(detector);
```

## Requirements
Before accepting this PR, there are a few things that need to be taken care of. Pre-building the Nan extension is great so people don't have to compile on their own, but it requires some configuring on the snowboy end of things if you want it to continue to work:
- You'll have to [enable Travis for snowboy](https://travis-ci.org/Kitt-AI/snowboy) if you want to leverage automatic building and publishing of the binary for Linux and OSX.
- I couldn't find any CI system that would work for ARM, so building and publishing for the Pi will have to be done manually :(
 - I have a bash script that you can use for this :)
- You'll have to add your own S3 bucket key and secret to [.travis.yml](https://github.com/evancohen/snowboy/blob/devel/.travis.yml#L32-L33) (or you can give me the secure strings and I can add them to this PR)
 - Instructions for that here: https://github.com/mapbox/node-pre-gyp#s3-hosting
 - Once that's set up, any commit message containing `[publish binary]` or `[republish binary]` (see examples in my commit history) will publish/republish the native binary for the version specified in `package.json`.
- Give me an email address to add to https://www.npmjs.com/package/snowboy so you can publish updates to the package.

🚀

### CODE REVIEW

**Comment by chenguoguo** [comment]
Thanks a lot for the hard work guys! I'll take a closer look tomorrow.

**Comment by xuchen** [comment]
looks good to me! I had to additionally do `npm install node-record-lpcm16` to install missing package.

**Comment by evancohen** [comment]
That might be worth including in the docs, but it's not a direct dependency of the module since you can pipe any arbitrary audio data (provided it's the correct format).

**Comment by chenguoguo** [comment]
I'll mention that in the docs. Will have to look into the pull request tomorrow, was working on some other stuff :-)

**Comment by chenguoguo** [comment]
I took a look at all the changed files, look good! I was thinking of squashing those commits into one, but on another thought, this contribution comes from two different developers, so let's keep those commits!

- I turned on Travis for Snowboy

- Could you share your bash script for raspberry pi?

- s3 bucket has been set up (I'll update after merging)
 - host: https://snowboy-release-node.s3-us-west-2.amazonaws.com/
 - secure strings generated using `travis encrypt node_pre_gyp_accessKeyId=xxx` and `travis encrypt node_pre_gyp_secretAccessKey=yyy`, where xxx and yyy are the actual keys. Let me know if this is not the right way.

- email: kittdotai@gmail.com

If things look good to you, I'm ready to merge! Thanks a million for the fantastic work @evancohen and @nekuz0r

**Comment by nekuz0r** [comment]
@chenguoguo @evancohen for me, commits can be squashed :)

**Comment by chenguoguo** [comment]
Thanks @nekuz0r ! I'll leave this to @evancohen . If we end up with squashing the commits, make sure you put @nekuz0r and @evancohen in the message so people know who contributed this commit :-)

**Comment by evancohen** [comment]
@chenguoguo I have the script sitting on my Pi at home, I can dig it up tonight. I'm also fine squashing the commits (or I can rebase on my side and squash everything into ~7 commits that make sense, instead of the 30 that we have).

Is that email address associated with an account on the npm site? I'm [having trouble adding it](https://cloud.githubusercontent.com/assets/1198365/18644377/0aa72e88-7e5e-11e6-86a0-a478720f287e.png) as a collaborator.

I did one final sanity check (https://github.com/evancohen/snowboy/commit/e7d547255aeaf52e29f7e7bf4e453565e2444587) incriminating the version number to [`1.0.10`](https://www.npmjs.com/package/snowboy), publishing the npm package, and running the builds on Travis to publish them:
https://travis-ci.org/evancohen/snowboy/builds/161123785

**Everything looks green on my end!**

There's one open question from @MexXxo on https://github.com/evancohen/snowboy/pull/1 about how we name the event handlers...
tl;dr If snowboy uses Voice Activity Detection then we shouldn't use an event named [`"noise"`](https://github.com/Kitt-AI/snowboy/pull/52/files#diff-9eb8f1e1192ca80a99f269ed604fd220R181) it should be called `"voice"`. Thoughts?

**Comment by chenguoguo** [comment]
@evancohen yes please rebase and squash if that's not too much trouble for you. Squashing to 7 sounds good!

Sorry about the email, please try again, the user name is kitt.ai.

For the VAD we have in Snowboy, it's more like silence v.s. non-silence. `voice` probably makes more sense than `noise`.

**Comment by evancohen** [comment]
I've rebased my branch and squashed both of our commits into larger changes, hopefully that's ok with @nekuz0r. 

@chenguoguo I've added that email as a collaborator to the npm package.

As for VAD, if it's actually "non-silence" then `noise` and `sound` would make sense - My concern for using `voice` is if someone assumes that any time that event fires someone is actually speaking (and not just moving a chair or something). Happy to make this change either way, I just want to ensure that it makes sense to people :)

**Comment by chenguoguo** [comment]
@evancohen thanks, looks very clean now!

Let's go with `sound` then, `noise` itself is not well defined, people have different opinions on ,e.g., treating `voice` as `noise` or treating `music` as `noise`. If you are going to make the change, I'll wait for that and merge. 

Good catch on the PR trigger! I turned it off.

**Comment by evancohen** [comment]
All set 🚀

**Comment by chenguoguo** [comment]
I updated `snowboy.umdl` so it works better in noisy environment; also added `alexa.umdl` although it doesn't work as well as snowboy. I think I'm ready to release v1.1.0 now.

@evancohen what extra steps does it take to publish the node module to npmjs.com? I've never done this by myself.

**Comment by nekuz0r** [comment]
@chenguoguo `npm publish` into the repository directory should enough

**Comment by evancohen** [comment]
That should do it. I would recommend publishing the binaries for 1.1.0 first if you haven't already.

**Comment by chenguoguo** [comment]
I did a test publish with 1.0.11 and everything seems to be working. I'll bump it to 1.1.0 tomorrow after publishing the binaries (don't have a Raspberry Pi by hand now).

**Comment by andreimavenhut** [comment]
Could this be tweaked to work in browser ? 
I haven't looked how you use the microphone or how you read the mdl / pmdl files.

**Comment by lynxaegon** [comment]
Could this be tweaked to work in browser ? 
I haven't looked how you use the microphone or how you read the mdl / pmdl files.

**Comment by arilotter** [comment]
@lynxaegon Not easily, since all the node module does is wrap a compiled library for some platform. Theoretically, it might be possible in the future for the kitt.ai team to compile the snowboy library for something like asm.js, but that part of the detection is closed source as of now.

**Comment by lynxaegon** [comment]
Ok, so i'm stuck with a python detector + a long poll request to detect a hotword.

**Comment by arilotter** [comment]
On the other hand, nothing's stopping you from running the node.js code on your server & streaming browser audio to it. Looks like that's what you're already doing with Python, though

**Comment by nekuz0r** [comment]
@lynxaegon 

@arilotter beat me on this, indeed you can stream your browser audio capture to a node server via websocket and get the detected keyword back via the same channel :)

**Comment by nick-jonas** [comment]
Is there anyway to import this into an electron project?  I get this error:

```Error: Cannot find module '/Users/njonas/Code/<project_dir>/node_modules/snowboy/lib/node/binding/Release/electron-v1.4-darwin-x64/snowboy.node'```

...when ```node-v48-darwin-x64``` is the one node binding that gets installed.

**Comment by evancohen** [comment]
@nick-jonas yup! It's for [Sonus](https://github.com/evancohen/sonus), but it'd be easy to adapt to use just for snowboy: https://github.com/evancohen/sonus-electron-boilerplate

**Comment by nick-jonas** [comment]
@evancohen that's great!  I get the same error though :(

```Uncaught Error: Cannot find module '/Users/njonas/Downloads/sonus-electron-boilerplate-master/node_modules/snowboy/lib/node/binding/Release/electron-v1.6-darwin-x64/snowboy.node'```

**Comment by evancohen** [comment]
@nick-jonas So you ran through the install steps in the README and the boilerplate didn't work?

It's possible that I inadvertently broke it when trying to fix something for ARM https://github.com/evancohen/sonus-electron-boilerplate/commit/3a5bd4ea239721164e1476112b62f6f6a5fa7919.

If that's the case then you can check out the [parent commit](https://github.com/evancohen/sonus-electron-boilerplate/commit/b06dfb2b706b2927156d933f74ad7b13796cfbdb) before that change and `rm -r node_modules` then go through the install steps again.

**Comment by nick-jonas** [comment]
This did the trick (from ^1.4):

```
"devDependencies": {
    "electron": "1.4.x"
  }
```

**Comment by evancohen** [comment]
What version do you have installed now? `npm ls electron`

**Comment by nick-jonas** [comment]
Electron version = 1.4.15.

Also, I'm not totally in the clear as I now get this error:

screen shot 2017-03-02 at 10 11 42 am

**Comment by evancohen** [comment]
@nick-jonas looks like you didn't install `sox` https://github.com/evancohen/sonus#for-macos

**Comment by nick-jonas** [comment]
ahh right that was it, thanks @evancohen

**Comment by Wqrld** [comment]
https://gist.github.com/Wqrld/a34ee45c8cf8edefad58eae5e33cdd7a it seems like the prebuilt files are inaccessible and compiling also doesnt work on ubuntu 16.04 Intel

**Comment by nekuz0r** [comment]
@Wqrld Your build seems to failed because the lib cblas can't be found on your system, install it and try again.
You might need to compile it by yourself.
http://www.netlib.org/blas/
